modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -278,48 +283,44 @@ private static void filterDeleteExceptionsAndRethrow(Exception e, IOException ex
-                    .flatMap(blob -> getDeleteTask(blob, blobContainerClient.getBlobAsyncClient(blob)), CONCURRENT_DELETES)
-                    .then()
-                    .block();
-            } catch (Exception e) {
-                filterDeleteExceptionsAndRethrow(e, new IOException("Unable to delete blobs"));
-            }
-        });
+        SocketAccess.doPrivilegedVoidException(() -> deleteListOfBlobs(getAzureBlobServiceClientClient(purpose), blobNames));
     }
 
-    private static Mono<Void> getDeleteTask(String blobName, BlobAsyncClient blobAsyncClient) {
-        return blobAsyncClient.delete()
-            // Ignore not found blobs, as it's possible that due to network errors a request
-            // for an already deleted blob is retried, causing an error.
-            .onErrorResume(
-                e -> e instanceof BlobStorageException blobStorageException && blobStorageException.getStatusCode() == 404,
-                throwable -> Mono.empty()
-            )
-            .onErrorMap(throwable -> new IOException("Error deleting blob " + blobName, throwable));
+    private void deleteListOfBlobs(AzureBlobServiceClient azureBlobServiceClient, Iterator<String> blobNames) throws IOException {
+        // We need to use a container-scoped BlobBatchClient, so the restype=container parameter
+        // is sent, and we can support all SAS token types
+        // See https://learn.microsoft.com/en-us/rest/api/storageservices/blob-batch?tabs=shared-access-signatures#authorization
+        final BlobBatchAsyncClient batchAsyncClient = new BlobBatchClientBuilder(
+            azureBlobServiceClient.getAsyncClient().getBlobContainerAsyncClient(container)
+        ).buildAsyncClient();
+        final List<Mono<Void>> batchResponses = new ArrayList<>();
+        while (blobNames.hasNext()) {
+            final BlobBatch currentBatch = batchAsyncClient.getBlobBatch();
+            int counter = 0;
+            while (counter < MAX_ELEMENTS_PER_BATCH && blobNames.hasNext()) {
+                currentBatch.deleteBlob(container, blobNames.next());
+                counter++;
+            }
+            batchResponses.add(batchAsyncClient.submitBatch(currentBatch));
+        }
+        try {
+            Flux.merge(batchResponses).collectList().block();
+        } catch (BlobBatchStorageException bbse) {
+            final Iterable<BlobStorageException> batchExceptions = bbse.getBatchExceptions();
+            for (BlobStorageException bse : batchExceptions) {
+                // If one of the requests failed with something other than a BLOB_NOT_FOUND, throw the encompassing exception
+                if (BlobErrorCode.BLOB_NOT_FOUND.equals(bse.getErrorCode()) == false) {
+                    throw new IOException("Failed to delete batch", bbse);
+                }
+            }
+        } catch (Exception e) {
+            throw new IOException("Unable to delete blobs", e);
```

Comment:
Should this call `filterDeleteExceptionsAndRethrow` or it is not necessary since no suppressed IOException due to the change of using blobBatch API?

Reply:
Very good pick up. I re-worked all of this. We need to
- Run all the batches to completion (some of the reactor utilities cancel the remaining ones when one fails)
- Selectively ignore batch failures that are only due to 404 responses

With that in mind, I changed to iterate over the response `Mono`s and `#block()` on each one, handling any exceptions individually. 

When you use something like `Mono#whenDelayError(monos)` it has the annoying habit of wrapping exceptions inconsistently depending on whether there is a single or multiple failures, also it adds in additional exceptions to the list of suppressed ones that contain reactor-specific info. The selective suppression of not found errors was much easier to read/write if I just did things without using the framework.

Happy to be schooled by someone who is familiar with the framework and knows better?

Reply:
Just need to check that that doesn't have the side-effect of executing them sequentially.

Reply:
Ugh it does.